server/src/main/java/org/apache/druid/server/http/CompactionResource.java

```diff
@@ -100,4 +114,31 @@ public Response getCompactionSnapshotForDataSource(
     }
     return Response.ok(ImmutableMap.of("latestStatus", snapshots)).build();
   }
+
+  @POST
+  @Path("/simulate")
+  @Consumes(MediaType.APPLICATION_JSON)
+  public Response simulateClusterCompactionConfigUpdate(
+      ClusterCompactionConfig updatePayload
+  )
+  {
+    if (compactionSupervisorsConfig.isEnabled()) {
+      return buildErrorResponseWhenRunningAsSupervisor();
+    }
+
+    return Response.ok().entity(
+        coordinator.simulateRunWithConfigUpdate(updatePayload)
+    ).build();
+  }
+
+  private Response buildErrorResponseWhenRunningAsSupervisor()
+  {
+    return Response.status(Response.Status.SERVICE_UNAVAILABLE).entity(
+        ImmutableMap.of(
+            "error",
+            "Compaction has been disabled on the Coordinator."
+            + " Use Overlord APIs to fetch compaction status."
+        )
+    ).build();
+  }
```

Comment:
If we just log a warning, we might return stale information from the coordinator.

Returning 302 would have sufficed if the path was the same for the two resources, but one path has prefix `/druid/coordinator/v1/compaction` and the other has `/druid/v1/indexer/compaction`.
The Druid clients `DruidLeaderClient` and `ServiceClientImpl` handle redirects but only update the service location and not the path.

Another option could be to just not do anything at the router, and simply use `OverlordClient` on the coordinator side and forward requests to the overlord if compaction supervisors are enabled. Let me know what you think.

Reply:
@abhishekagarwal87 , to deal with the issue with rolling upgrades, I decide to do this instead.
- The leader Overlord is the single source of truth on whether the feature is enabled.
- API `/druid/indexer/v1/compaction/isSupervisorEnabled` can be used by the Coordinator to determine if the feature is enabled.
- If enabled, Coordinator redirects the compaction status APIs to the leader Overlord using the `OverlordClient`.
- Router does not perform any redirecting between Overlord and Coordinator.
- __Overlord does not redirect to Coordinator.__ When the feature is disabled, Overlord simply returns a 504 (service unavailable) with an appropriate error message. This is to keep things simple and to avoid getting into an infinite redirect situation.

The only reason we need the redirect is that a client (like web-console) using the Coordinator compaction status APIs continues to work seamlessly when we toggle the feature on or off. With a client explicitly using the new Overlord APIs, we don't have the burden of backward compatibility and thus no redirect is required.